### PR TITLE
A method added to Colormap classes to reverse the colormap

### PR DESCRIPTION
**Pull request by kjartankg**
This implements a solution for issue #4271

* method added to both `ListedColormap` and `LinearSegmentedColormap`
* The parent class `Colormap` raises `NotImplementedError`
* A test was added to `test_colors.py`

This PR succeeds PR #5899. Following is the initial comment copy/pasted from the previous PR:

A few things that I ran into while implementing this:

* Functions to reverse Colormap data already existed in [matplotlib.cm](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/cm.py) but they are inconvenient for a user. After seeing that I changed my implementation to resemble them more. Perhaps the old functions should eventually be deprecated?
* The documentation for [LinearSegmentedColormap](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/colors.py#L669) does not mention that input data can be functions. It is mentioned briefly in the doc for [makeMappingArray](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/colors.py#L429) but should probably be mentioned in the doc for LinearSegmentedColormap.
* The data is specified as "list of tuples" [LinearSegmentedColormap](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/colors.py#L669) and [ListedColormap](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/colors.py#L788) "a list of matplotlib color specifications, or an equivalent Nx3 or Nx4 floating point array (*N* rgb or rgba values)". I would suggest that this would be changed to *array_like* and that the data is then copied into a numpy array for internal use.

figure_1
figure_2
figure_3
figure_4
figure_5
figure_6

### CODE REVIEW

**Comment by kjartankg** [comment]
Looking at the Tracis CI test, I understand the pep8 failure (missing a blank line in the test_colors.py) but the other failure I don't understand. Can someone explain that to me? It seems to be unrelated to what I've done in this PR.

**Comment by mdboom** [comment]
The other failure is testing against Numpy nightlies, which is known to be broken.  It's an optional test, though -- it doesn't cause the whole thing to fail when it fails.

**Comment by mdboom** [comment]
Cc: @efiring

**Comment by efiring** [comment]
Based on a quick look, I think this looks promising.

File: lib/matplotlib/colors.py [hunk 1]
```diff
@@ -665,6 +665,30 @@ def _resample(self, lutsize):
         """
         raise NotImplementedError()
 
+    def reversed(self, name=None):
+        """
+        Make a reversed instance of the Colormap.
+
+        NOTE: Function not implemented for base class.
```
Comment (tacaswell) [inline lib/matplotlib/colors.py]: Use `.. note ::` and sphinx will render this better.

File: lib/matplotlib/colors.py [hunk 2]
```diff
@@ -665,6 +665,30 @@ def _resample(self, lutsize):
...
+
+        Parameters
+        ----------
+        name : str, optional
+            The name for the reversed colormap. If it's None the
+            name will be the name of the parent colormap + "_r".
+
+        Raises
```
Comment (tacaswell) [inline lib/matplotlib/colors.py]: Don't need this section.

**Comment by tacaswell** [comment]
Can you also add a note in `whats_new` https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new ?

**Comment by tacaswell** [comment]
@kjartankg Thanks!  I think this is your first contribution to mpl, congratulations!  Hope we hear from you again.

**Comment by kjartankg** [comment]
Excellent, thanks! I will most likely have a look at some more things. @tacaswell if there's anything specific that needs to be worked on feel free to point me in that direction. Otherwise I will just browse the issues again.

**Comment by tacaswell** [comment]
Browsing the issues / fixing itches you encounter day-to-day is the best place to start.

**Pull request merged**